File: packages/cli-py/planforge/utils/plan_files.py
"""Parse '## Files Likely to Change' section from plan markdown and return file paths/globs."""

import re

SECTION_HEADING = "## Files Likely to Change"
LIST_ITEM_RE = re.compile(r"^[-*]\s+(?:`([^`]+)`|(.+))$")


def parse_files_from_plan(plan_content: str | None) -> list[str]:
    """Extract file paths from the 'Files Likely to Change' section of a plan document.
    Returns unique non-empty paths (backticks stripped). Globs are included as strings.
    """
    if not (plan_content and plan_content.strip()):
        return []
    lines = plan_content.splitlines()
    in_section = False
    seen: set[str] = set()
    paths: list[str] = []
    for line in lines:
        trimmed = line.strip()
        if trimmed.startswith("## "):
            if trimmed.lower().startswith(SECTION_HEADING.lower()):
                in_section = True
            else:
                in_section = False
            continue
        if not in_section:
            continue
        m = LIST_ITEM_RE.match(trimmed)
        if not m:
            continue
        raw = (m.group(1) or m.group(2) or "").strip()
        path = raw.lstrip("/").replace("\\", "/")
        if not path or ".." in path or path in seen:
            continue
        seen.add(path)
        paths.append(path)
    return paths

File: packages/cli-js/src/utils/plan-files.ts
/**
 * Parse "## Files Likely to Change" section from plan markdown and return file paths/globs.
 * Lines like "- path" or "- `path`" are extracted; globs are kept as-is for "files to focus on" text.
 */

const SECTION_HEADING = "## Files Likely to Change";
const LIST_ITEM_RE = /^[-*]\s+(?:`([^`]+)`|(.+))$/;

/**
 * Extract file paths from the "Files Likely to Change" section of a plan document.
 * Returns unique non-empty paths (backticks stripped). Globs are included as strings.
 */
export function parseFilesFromPlan(planContent: string | undefined): string[] {
  if (!planContent?.trim()) return [];

  const lines = planContent.split(/\r?\n/);
  let inSection = false;
  const seen = new Set<string>();
  const paths: string[] = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.startsWith("## ")) {
      if (trimmed.toLowerCase().startsWith(SECTION_HEADING.toLowerCase())) {
        inSection = true;
        continue;
      }
      inSection = false;
      continue;
    }
    if (!inSection) continue;
    const m = trimmed.match(LIST_ITEM_RE);
    if (!m) continue;
    const path = (m[1] ?? m[2] ?? "").trim().replace(/^\/+/, "").replace(/\\/g, "/");
    if (!path || path.includes("..")) continue;
    if (seen.has(path)) continue;
    seen.add(path);
    paths.push(path);
  }

  return paths;
}
